quedex_api/user_stream.py

```diff
@@ -0,0 +1,302 @@
+      }
+    """
+    pass
+
+  def on_error(self, error):
+    """
+    Called when an error with market stream occurs (data parsing, signature verification, webosocket
+    error). This means a serious problem, which should be investigated (cf. on_disconnect).
+
+    :type error: subtype of Exception
+    """
+    pass
+
+  def on_disconnect(self, message):
+    """
+    Called when market stream disconnects cleanly (exchange going down for maintenance, network
+    problem, etc.). The client should reconnect in such a case.
+
+    :param message: string message with reason of the disconnect
+    """
+    pass
+
+
+class UserStream(object):
+  def __init__(self, exchange, trader, nonce_group=5):
+    super(UserStream, self).__init__()
+    self.send_message = None
+    self.user_stream_url = exchange.user_stream_url
+
+    self._quedex_key = exchange.public_key
+    self._trader_key = trader.private_key
+    self._account_id = trader.account_id
+
+    self._listeners = []
+    self._nonce_group = nonce_group
+    self._nonce = None
+    self._initialized = False
+
+  def add_listener(self, listener):
+    self._listeners.append(listener)
+
+  def remove_listener(self, listener):
+    self._listeners.remove(listener)
+
+  def place_order(self, place_order_command):
+    """
+    :param place_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id unique among orders>,
+        "instrument_id": "<string id of the instrument>",
+        "order_type": "limit",
+        "limit_price": "<decimal as string>",
+        "side": "buy"/"sell",
+        "quantity": <integer>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    place_order_command['type'] = 'place_order'
+    self._encrypt_send(self._set_nonce_account_id(place_order_command))
+
+  def cancel_order(self, cancel_order_command):
+    """
+    :param cancel_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id of the order to cancel>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    cancel_order_command['type'] = 'cancel_order'
+    self._encrypt_send(self._set_nonce_account_id(cancel_order_command))
+
+  def modify_order(self, modify_order_command):
+    """
+    :param modify_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id of the order to modify>,
+        "new_limit_price": "<decimal as string>",
+        "new_quantity": <integer>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    modify_order_command['type'] = 'modify_order'
+    self._encrypt_send(self._set_nonce_account_id(modify_order_command))
+
+  def batch(self, order_commands):
```

Comment:
I definitely think that batching should be explicit. This is the main functional feature of batching - having "batch orders":

- a market maker has the ability to control that all his order are changed "at once"/"atomically" without anyone's orders interleaving (moving both sides of an order book simultaneously, or even multiple instruments simultaneously) - this was the main rationale for introducing batches
- provides means for other "atomic" operations, like Immediate or Cancel, which have to explicitly implemented by the trader to make sense

The fact that it also provides much faster execution, especially with our PGP encryption is a great nonfunctional benefit, but it's actually a side-effect.

Additional, automatic batching of commands in a single reactor "tick" might be interesting (we could try with Twisted's `callLater` etc., if they provide `setImmediate` or similar semantics), but may be tricky and is not a must-have for now, IMO. As far as aggregating among many listeners - I don't think that would be that beneficial - I expect most people to attach only a single listener.

I personally prefer (and have seen some people also do) rather `batch([commands])` methods when implementing strategies, because (especially as a market maker) you usually already manage a some collection of orders which you transform. I think that's a matter of taste.

I will add `begin_batch` and `send_batch` as that's trivial and leave `batch`. WDYT?